homeassistant/components/vacuum/ecovacs.py

```diff
@@ -0,0 +1,195 @@
+"""
+Support for Ecovacs Ecovacs Vaccums.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/vacuum.neato/
+"""
+import logging
+
+from homeassistant.components.vacuum import (
+    VacuumDevice, SUPPORT_BATTERY, SUPPORT_RETURN_HOME, SUPPORT_CLEAN_SPOT,
+    SUPPORT_STATUS, SUPPORT_STOP, SUPPORT_TURN_OFF, SUPPORT_TURN_ON,
+    SUPPORT_LOCATE, SUPPORT_FAN_SPEED, SUPPORT_SEND_COMMAND, )
+from homeassistant.components.ecovacs import (
+    ECOVACS_DEVICES)
+from homeassistant.helpers.icon import icon_for_battery_level
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['ecovacs']
+
+SUPPORT_ECOVACS = (
+    SUPPORT_BATTERY | SUPPORT_RETURN_HOME | SUPPORT_CLEAN_SPOT |
+    SUPPORT_STOP | SUPPORT_TURN_OFF | SUPPORT_TURN_ON | SUPPORT_LOCATE |
+    SUPPORT_STATUS | SUPPORT_SEND_COMMAND | SUPPORT_FAN_SPEED)
+
+ATTR_ERROR = 'error'
+ATTR_COMPONENT_PREFIX = 'component_'
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Ecovacs vacuums."""
+    vacuums = []
+    for device in hass.data[ECOVACS_DEVICES]:
+        vacuums.append(EcovacsVacuum(device))
+    _LOGGER.debug("Adding Ecovacs Vacuums to Hass: %s", vacuums)
+    add_devices(vacuums, True)
+
+
+class EcovacsVacuum(VacuumDevice):
+    """Ecovacs Vacuums such as Deebot."""
+
+    def __init__(self, device):
+        """Initialize the Ecovacs Vacuum."""
+        self.device = device
+        self.device.connect_and_wait_until_ready()
```

Comment:
I/O should be avoided in `__init__` as it blocks the event loop, better connect inside update or before adding the device.

Reply:
In this case it doesn't block the event loop since we instantiate the entity in `setup_platform` which is run in the thread pool.

But I like to avoid I/O in init on a general bases anyway.

Reply:
Would the preferred implementation to be to connect to the device in `setup_platform` and then pass the already-connected `device` object in to the `__init__`? That's an easy change; just want to confirm that's the best practice here.

Reply:
If `connect_and_wait_until_ready()` only depends on the API being available, moving it to `setup_platform` is fine.If the call can fail if vacuum is not available when HA starts, it would be better to allow it to be brought on-line asynchronously in `EcovacsVacuum.update`. You can use `add_devices(..., update_before_add=True)` if important attributes like the name are not available until connected.